### PR TITLE
Set application/x-ndjson content type on bulk requests.

### LINKED ISSUE #609: Set `application/x-ndjson` content type on bulk requests
As of elastic 5.x, requests to `/_bulk` should set `Content-Type` to `application/x-ndjson`. If not, elastic logs a warning. It looks like this library defaults to `application/json`. To fix, I'm thinking we should accept an optional dict of headers at `https://github.com/elastic/elasticsearch-py/blob/master/elasticsearch/connection/http_urllib3.py#L94`. If that sounds reasonable, I'd be happy to submit a patch.

cc @cnelson @wjwoodson

### PR DESCRIPTION
**Pull request by jmcarp**
Resolves #609

### CODE REVIEW

File: elasticsearch/connection/http_urllib3.py [hunk 1]
```diff
@@ -111,6 +111,8 @@ def perform_request(self, method, url, params=None, body=None, timeout=None, ign
             if not isinstance(method, str):
                 method = method.encode('utf-8')
 
+            request_headers = dict(self.headers)
+            request_headers.update(headers or {})
```
Comment (HonzaKral) [inline elasticsearch/connection/http_urllib3.py]: It would be nice if this block only ran `if headers:` to avoid copying the `self.headers` dict every time when it is only needed for the bulk-like bodies

File: elasticsearch/client/__init__.py [hunk 2]
```diff
@@ -1164,7 +1164,8 @@ def bulk(self, body, index=None, doc_type=None, params=None):
         if body in SKIP_IN_PATH:
             raise ValueError("Empty value passed for a required argument 'body'.")
         return self.transport.perform_request('POST', _make_path(index,
-            doc_type, '_bulk'), params=params, body=self._bulk_body(body))
+            doc_type, '_bulk'), params=params, body=self._bulk_body(body),
+            headers={'Content-Type': 'application/x-ndjson'})
```
Comment (HonzaKral) [inline elasticsearch/client/__init__.py]: we use lowercase for all headers to avoid conflicts - see https://github.com/elastic/elasticsearch-py/blob/master/elasticsearch/connection/http_urllib3.py#L59-L62

A test is needed to make sure the headers are merged properly

Comment (HonzaKral) [inline elasticsearch/client/__init__.py]: There are other APIs using the `_bulk_body` like `msearch` and `mpercolate`

File: elasticsearch/transport.py [hunk 3]
```diff
@@ -269,6 +269,8 @@ def perform_request(self, method, url, params=None, body=None):
 
         :arg method: HTTP method to use
         :arg url: absolute url (without host) to target
+        :arg headers: dictionary of headers, will be handed over to the
+            underlying :class:`~elasticsearch.Connection` class for serialization
```
Comment (HonzaKral) [inline elasticsearch/transport.py]: no serialization happen to the headers, we need to set expectations correctly

**Comment by davidt99** [comment]
@jmcarp Any plan to fix the review comments? Most of the driver requests (if not all) doesn't state content type, resulting warning logs in elastic 5.5.
The header support you wrote will be allow adding content type to all the request, thus fixing the warning.

**Comment by HonzaKral** [comment]
@davidt99 All requests are anotated with content-type accepted by elasticsearch since d862c8a (released in `5.2`). You can also set headers globally by specifying `headers={'custom': 'header'}` when instantiating `Elasticsearch`.

**Comment by davidt99** [comment]
@HonzaKral Oh, I was using 5.1 and looked for similar issue. Thanks.

**Comment by jmcarp** [comment]
I cleaned this up based on your suggestions @HonzaKral; let me know if this needs more revisions. I'm not sure why the tests are failing, but it looks like they're failing in the same way as `master`.

File: elasticsearch/client/__init__.py [hunk 4]
```diff
@@ -1159,7 +1160,158 @@ def msearch(self, body, index=None, doc_type=None, params=None):
...
+        :arg routing: A comma-separated list of specific routing values
+        :arg version: Explicit version number for concurrency control
+        :arg version_type: Specific version type, valid choices are: 'internal',
+            'external', 'external_gte', 'force'
+        """
+        for param in (index, doc_type):
+            if param in SKIP_IN_PATH:
+                raise ValueError("Empty value passed for a required argument.")
+        return self.transport.perform_request('GET', _make_path(index,
+            doc_type, id, '_percolate', 'count'), params=params, body=body)
```
Comment (HonzaKral) [inline elasticsearch/client/__init__.py]: several APIs have been removed from `master` as they have been removed in elasticsearch `6.0`. Either please remove them from the patch or base the patch on `5.x`

File: elasticsearch/client/__init__.py [hunk 5]
```diff
@@ -1387,4 +1540,3 @@ def field_caps(self, index=None, body=None, params=None):
         """
         return self.transport.perform_request('GET', _make_path(index,
             '_field_caps'), params=params, body=body)
-
```
Comment (HonzaKral) [inline elasticsearch/client/__init__.py]: please configure your editor not to make changes like these. It makes the PR hard to review when they are unrelated stylistic changes in unrelated places.

File: elasticsearch/connection/base.py [hunk 6]
```diff
@@ -123,5 +123,3 @@ def _raise_error(self, status_code, raw_data):
             logger.warning('Undecodable raw error response from server: %s', err)
 
         raise HTTP_EXCEPTIONS.get(status_code, TransportError)(status_code, error_message, additional_info)
-
-
```
Comment (HonzaKral) [inline elasticsearch/connection/base.py]: please configure your editor not to make changes like these. It makes the PR hard to review when they are unrelated stylistic changes in unrelated places.

File: elasticsearch/transport.py [hunk 7]
```diff
@@ -292,7 +294,7 @@ def perform_request(self, method, url, params=None, body=None):
 
         if body is not None:
             try:
-                body = body.encode('utf-8', 'surrogatepass')            
+                body = body.encode('utf-8', 'surrogatepass')
```
Comment (HonzaKral) [inline elasticsearch/transport.py]: please configure your editor not to make changes like these. It makes the PR hard to review when they are unrelated stylistic changes in unrelated places.

File: test_elasticsearch/test_server/test_helpers.py [hunk 8]
```diff
@@ -16,7 +16,7 @@ def bulk(self, *args, **kwargs):
         self._called += 1
         if self._called in self._fail_at:
             raise self._fail_with
-        return  self.client.bulk(*args, **kwargs)
+        return self.client.bulk(*args, **kwargs)
```
Comment (HonzaKral) [inline test_elasticsearch/test_server/test_helpers.py]: please configure your editor not to make changes like these. It makes the PR hard to review when they are unrelated stylistic changes in unrelated places.

**Comment by jmcarp** [comment]
I rebased on master and dropped the deprecated methods, and reverted whitespace changes. This is ready for another look when you have time @HonzaKral.

**Comment by fxdgear** [comment]
This LGTM. Can you please rebase off master and push again? I had to change the version of ES that travis uses.

**Comment by jmcarp** [comment]
Done @fxdgear.

**Comment by fxdgear** [comment]
Great thanks @jmcarp
